node master will now use SIGKILL on worker after timeout resolves #605

jsnoble opened the pull request:
NOTE CONFIG SETTING CHANGES:

- changed `network_timeout` to `action_timeout`
- added `network_latency_buffer` (defaults to 15s) which combines with action_timeout to make the final network_timeout

- can now add a a timeout parameter to stop job which will take the place of action_timeout, note that `network_latency_buffer` will still be added to it for final timeout

```
curl -XPOST localhost:5678/jobs/8b9596db-db1b-4db4-8957-b2bdc4625335/_stop?timeout=120000
```

kstaken commented on lib/cluster/node_master.js:
```diff
@@ -187,8 +187,10 @@ module.exports = function module(context) {
+            const filterFn = worker => worker.ex_id === data.ex_id;
+            let stopTime = data.timeout || config.network_timeout;
+
+            messageWorkers(data, { message: 'worker:shutdownzzz' }, filterFn);
```
worker:shutdownzzz?

kstaken commented:
Had a scenario today where workers were refusing to exit and decided to try this branch. This did not work to kill the workers.

jsnoble commented:
Sorry about that, just changed it back. I changed the message name so that it would not be recoginzed so that I could verify that the process would be killed after the alloated time

kstaken commented:
This doesn't have any effect unless a stop job is called through the API. And even then it doesn't work. I tried logging stopTime on line 215 of node_master and it is NaN for me.

This needs to work in all scenarios where workers should be stopped.

kstaken commented:
```let stopTime = data.timeout || config.network_timeout;```

Both of those options show up as undefined for me.

jsnoble commented:
Ok, fixed the wrong reference, it should say config.action_timeout

kstaken commented:
This PR is not ready. 
* Worker shutdown behavior should be handled in a consistent manner for all cases not just coming through the API. Hitting ctrl-c, normal termination of the node_master, remove worker and stopping the job all need to guarantee workers exit.
* The API still times out and provides no result before workers exit with default timeouts.
* network_timeout has been renamed but appears to still be referenced in: ./cluster/services/cluster/backends/native.js:            }, context.sysconfig.teraslice.network_timeout);

kstaken commented:
Even when the timeout triggers and kill is called it still doesn't really kill the workers. They are removed from cluster state so don't show up under /txt/workers but the processes are still there at the OS level.

kstaken commented:
I was able to get it to actually kill the processes with this.

```
    function killWorkers(filterFn) {
        const spawn = require( 'child_process' ).spawnSync; <-- obviously wouldn't belong here.
        const workers = context.cluster.workers;
        const allWorkersForJob = _.filter(workers, filterFn);

        allWorkersForJob.forEach(worker => {
            spawn('kill', ['-9', worker.process.pid]);
        });
    }
```

godber commented:
This one is still a WIP right?

kstaken commented:
Manually confirmed ctrl-c, API _stop and API _remove workers will eventually kill stuck workers. 

The default timeout for _stop is longer than the default timeout for curl so you'll get a curl timeout before the workers are actually killed. Also it appeared the timeout on _remove and ctrl-c was shorter than for _stop.

kstaken commented:
Integration tests pass.
